src/xmipp/libraries/reconstruction_cuda/cuda_volume_restoration_kernels.cpp

```diff
@@ -0,0 +1,267 @@
+
+	normalizeForFFTKernel<<<dimGrid, dimBlock>>>(d_V1, d_V2, volume_size, static_cast<T>(1.0) / volume_size);
+	gpuErrchk(cudaPeekAtLastError());
+}
+
+template< typename T >
+void VolumeRestorationKernels<T>::normalizeForFFT(T* d_V1, size_t volume_size) {
+	unsigned block_size = 256;
+
+	dim3 dimBlock{ block_size };
+	dim3 dimGrid{ static_cast<unsigned>( ceil( static_cast<double>(volume_size) / dimBlock.x) ) };
+
+	normalizeForFFTKernel<<<dimGrid, dimBlock>>>(d_V1, volume_size, static_cast<T>(1.0) / volume_size);
+	gpuErrchk(cudaPeekAtLastError());
+}
+
+template< typename T >
+void VolumeRestorationKernels<T>::restorationSigmaCostError(T& error, const std::complex<T>* _d_fVol, const std::complex<T>* _d_fV1, const std::complex<T>* _d_fV2, const T* __restrict__ d_R2, T K1, T K2, size_t fourier_size) {
+	const T inv_size = 1.0 / (2 * fourier_size);
+
+	const vec2_type<T>* __restrict__ d_fVol = (vec2_type<T>*)_d_fVol;
+	const vec2_type<T>* __restrict__ d_fV1 = (vec2_type<T>*)_d_fV1;
+	const vec2_type<T>* __restrict__ d_fV2 = (vec2_type<T>*)_d_fV2;
+
+	auto error_func = [=] __device__ (int n) {
+		const T R2n = d_R2[n];
+		if (R2n <= 0.25) {
+			const T H1 = exp(K1 * R2n);
+			const T H2 = exp(K2 * R2n);
+
+			const T diff1_x = (d_fVol[n].x*H1 - d_fV1[n].x) * inv_size;
+			const T diff1_y = (d_fVol[n].y*H1 - d_fV1[n].y) * inv_size;
+
+			const T diff2_x = (d_fVol[n].x*H2 - d_fV2[n].x) * inv_size;
+			const T diff2_y = (d_fVol[n].y*H2 - d_fV2[n].y) * inv_size;
+
+			return sqrt(diff1_x*diff1_x + diff1_y*diff1_y) + sqrt(diff2_x*diff2_x + diff2_y*diff2_y);
+		}
+
+		return static_cast<T>(0.0);
+	};
+
+	error = thrust::transform_reduce(thrust::device, thrust::counting_iterator<int>(0), thrust::counting_iterator<int>(fourier_size), error_func, static_cast<T>(0), thrust::plus<T>());
+}
+
+template< typename T >
+void VolumeRestorationKernels<T>::computeDiffAndAverage(const T* __restrict__ d_V1, const T* __restrict__ d_V2, T* __restrict__ d_S, T* __restrict__ d_N, size_t volume_size) {
+	auto k = [=] __device__ (int n) {
+		d_N[n] = d_V1[n] - d_V2[n];
+		d_S[n] = (d_V1[n] + d_V2[n]) * static_cast<T>(0.5);
+	};
+
+	thrust::for_each_n(thrust::device, thrust::counting_iterator<int>(0), volume_size, k);
+}
+
+template< typename T >
+std::pair<T, T> normAvgStd(T avg, T std, size_t size) {
+	avg /= size;
+
+	if (size > 1) {
+		std = std / size - avg * avg;
+		std *= static_cast<T>(size) / (size - 1);
+		std = sqrt(abs(std));
+	} else {
+		std = 0;
+	}
+
+	return { avg, std };
+}
+
+template< typename T >
+std::pair<T, T> VolumeRestorationKernels<T>::computeAvgStd(const T* __restrict__ d_N, size_t volume_size) {
+	const T avg = thrust::reduce(thrust::device, d_N, d_N + volume_size);
+
+	auto square_kernel = [=] __device__ (T x) {
+		return x * x;
+	};
+
+	const T std = thrust::transform_reduce(thrust::device, d_N, d_N + volume_size, square_kernel, static_cast<T>(0), thrust::plus<T>());
```

Comment:
I guess we can join these two kernels, if we need a bit more performance

Reply:
I don't see an easy way how to join them, unless there is a touple/pair that works on device.